### DISCUSSION TITLE
Is it possible to deactivate imgui in a specific build configuration ?

### DISCUSSION THREAD

**Discussion by Tichau** [pkdawson/imgui-godot#62] [Q&A] [answered]
Hello,

Great work on this bindings, it works very well. I use imgui to make in-game tools in my game. I was wondering if it is possible to completely remove it from a specific export preset (for example using feature tags) in order to not deliver content that is not used by the player ?

Thanks,

**Comment by pkdawson** [comment]
First, the easy part: in the export dialog, go to Resources and add `addons/imgui-godot/*` to the exclude filter. That should remove the plugin completely from the export.

The tricky part is disabling your code that uses ImGui. For C#, you could add something like this to your .csproj:
```csproj
  <PropertyGroup>
    <DefineConstants Condition="'$(Configuration)'=='Debug' And
        $(DefineConstants.Contains('GODOT_PC'))">$(DefineConstants);IMGUI</DefineConstants>
  </PropertyGroup>
```

And then conditionally enable your code that uses ImGui, like:
```csharp
#if IMGUI
ImGui.Begin("my window");
// ...
ImGui.End();
#endif
```

For C++, same thing. Use the preprocessor.

For GDScript, I don't have any great answers. You might be able to put all your tool scripts in one directory, and add that directory to the export exclusion filter.

Let me know if you have any specific difficulty. The issue has come up a few times, so I'd like to document it better and maybe add an example project to show "best practices" for exporting.

**Reply by Tichau** [reply]
Thanks ! 
It works well for Debug and ExportRelease configuration.

But I also want to have my imgui debug ui in ExportDebug builds (build made with --export-debug flag or with `Export With Debug` checked using UI). Do I need to create a separated export preset or is there a way to change resource to exclude depending on if it's a Debug build or not ?

**Reply by pkdawson** [reply]
Yeah I think your only option there is to make another preset which doesn't exclude addons/imgui-godot

**Comment by Tichau** [comment]
Here is a recap to document the way I've done for my project following the advices of @pkdawson.

I'm using C# version of Godot so I edited the csproj to add a new define when I want imgui to be activated:
```xml
<PropertyGroup>
    <DefineConstants Condition="$(DefineConstants.Contains('DEBUG')) And $(DefineConstants.Contains('GODOT_PC'))">$(DefineConstants);IMGUI</DefineConstants>
</PropertyGroup>
```

then, I used this define to conditionally add a reference on ImGui.NET or to remove the addon code from the project:
```xml
<ItemGroup Condition="$(DefineConstants.Contains('IMGUI'))">
    <PackageReference Include="ImGui.NET" Version="1.90.8.1" />
</ItemGroup>
<ItemGroup Condition="!$(DefineConstants.Contains('IMGUI'))">
    <Compile Remove="addons\imgui-godot\**" />
</ItemGroup>
```

I changed the ImGuiRoot option depending on the release tag to un-reference the associated scene:
```
[autoload]
ImGuiRoot="*res://addons/imgui-godot/data/ImGuiRoot.tscn"
ImGuiRoot.release="res://addons/imgui-godot/data/ImGuiRoot.tscn"
```

I excluded the imgui-godot addon from the resources defined in the export preset.
`exclude_filter="addons/imgui-godot/*"`

To be able to make a "debug" build, I duplicated the export preset without this exclusion.

Finally, I used the IMGUI define to include ImGUI uses: 
```csharp
#if IMGUI
ImGui.Begin("my window");
// ...
ImGui.End();
#endif
```

Thanks again for the help

**Comment by Tichau** [comment, marked as answer]
Hi again,

So, I spoke to fast, I did not tested enough what I've done ^^. My previous integration was only working for exported build. ImGui was not working when testing in editor. 

I tried different approaches, it is quite difficult to add new define constants in a godot project (so it works well in editor too). So I used Godot [feature tags](https://docs.godotengine.org/en/4.2/tutorials/export/feature_tags.html) that seems to exist to do this kind of stuff. Sadly for now feature tags cannot be set in editor (but it will starting from Godot 4.3, see issue [#373](https://github.com/godotengine/godot-proposals/issues/373)).
So for now, I implemented it using negative logic, imgui is active by default, then I use a feature tag to deactivate it.

Here is what I've done:

* I added the following choose statement in the project csproj:
```xml
<Choose>
    <When Condition="$(DefineConstants.Contains('GODOT_FEATURE_NO_IMGUI'))">
        <ItemGroup>
            <Compile Remove="addons\imgui-godot\**" />
        </ItemGroup>
    </When>
    <Otherwise>
        <ItemGroup>
            <PackageReference Include="ImGui.NET" Version="1.90.8.1" />
        </ItemGroup> 
    </Otherwise>
</Choose>
```

* I edited the release export preset to exclude the imgui-godot addon from the resources and to add a custom feature flag
```
custom_features="no_imgui"
exclude_filter="addons/imgui-godot/*"
```

To be able to make a "debug" build that use imgui, I duplicated the export preset without this exclusion and flag.

* Finally, I used the !GODOT_FEATURE_NO_IMGUI define to include ImGUI uses:

#if !GODOT_FEATURE_NO_IMGUI
ImGui.Begin("my window");
// ...
ImGui.End();
#endif

This integration works well in export and in editor. I think I'll change to positive logic when 4.3 will be out, but this is fine that way and it's also allow to override some project settings depending on this new flag.

**Reply by Tichau** [reply]
I think I mixed up things during my tests. The GODOT_FEATURE_NO_IMGUI define constant is now added by Godot. It's in fact only a PR for now (https://github.com/godotengine/godot/pull/53920).
For now it needs to be checked at runtime using `OS.HasFeature("no_imgui")`

**Comment by pkdawson** [comment]
Thanks for sharing your work. I borrowed some of your ideas for the [sample csproj](https://github.com/pkdawson/imgui-godot/blob/master/doc/examples/CSharpGameProject/CSharpGameProject.csproj), which works well in my testing with imgui-godot v6.0.0.

No feature flags or exclusion filters necessary, you can just use the [new settings](https://github.com/pkdawson/imgui-godot/wiki/Export) in each export preset.

And to fix the using directives for release export, you can just do:

```csharp
#if !IMGUI
namespace ImGuiNET { }
namespace ImGuiGodot { }
#endif
```
